### PR TITLE
[core] Warnings and a bit of cleanup of the core.

### PR DESCRIPTION
**Pull request by maceligueta**
I worked a bit on getting rid of some annoying warnings.
Some of them are trivial (unused variables) but others are a bit more delicate, please take a look.
One of those fixes comes from a valgrind notification when running the cpp tests (commits are clearly separated).
The repeated registry of variables when loading the Kratos Kernel was also removed.

### CODE REVIEW

File: kratos/python/add_constitutive_law_to_python.cpp [hunk 1]
```diff
@@ -91,10 +91,31 @@ void DeprecatedInitializeSolutionStep( ConstitutiveLaw& rThisConstitutiveLaw,
...
                                                 rShapeFunctionsValues,
                                                 rCurrentProcessInfo);
+
+    #if defined(__clang__)
+    #pragma clang diagnostic pop
+    #elif defined(__GNUC__) || defined(__GNUG__)
+    #pragma GCC diagnostic pop
+    #elif defined(_MSC_VER)
+    #pragma warning(pop)
+    #endif
```
Comment (philbucher) [inline kratos/python/add_constitutive_law_to_python.cpp]: why is this needed?

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h [hunk 2]
```diff
@@ -236,7 +238,7 @@ class BuilderAndSolver
 
         mReshapeMatrixFlag = false; //by default the matrix is shaped just once
         //		mVectorsAreInitialized = false;
-
+        mCalculateReactionsFlag = false;
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: I think it would be better to initialize the member to False rather than doing it separately in every constructor
=> If someone adds a new constructor the same problem might appear again

File: kratos/sources/kernel.cpp [hunk 3]
```diff
@@ -25,7 +25,7 @@ Kernel::Kernel() : mpKratosCoreApplication(Kratos::make_shared<KratosApplication
                     << "           Multi-Physics " << KRATOS_VERSION << std::endl;
 
     if (!IsImported("KratosMultiphysics")) {
-        mpKratosCoreApplication->RegisterVariables();
+      //mpKratosCoreApplication->RegisterVariables();
```
Comment (philbucher) [inline kratos/sources/kernel.cpp]: please remove code, do not comment
This is really confusing if someone else looks at this (thinking if this is intentional or debugging leftovers)
Esp in those parts of the code

The old state can always be recovered through the git-history

**Review by philbucher — DISMISSED** [review]
please remove code, do not comment it

File: kratos/python/add_constitutive_law_to_python.cpp (hunk 1 above)
Comment (maceligueta) [reply kratos/python/add_constitutive_law_to_python.cpp]: Right now, when we compile the core, everyone gets a warning saying that the method InitializeSolutionStep is deprecated. I think it's not nice to show this warning to everyone, specially to those who don't use this method or don't care about it. However, this method is exported to Python, so it means that the core contains a call to this deprecated method. 
For calls from Python, a KRATOS_WARNING appears in runtime (OK). For calls from c++ (applications) a compiler warning appears (OK). But for this single call from the core, I am trying to hide the warning, and this is what this implementation does, it deactivates the 'deprecated' warning just for one line of the core.

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h (hunk 2 above)
Comment (maceligueta) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: Should I do this for the other variables as well? I just followed the style of the file.

Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: > Should I do this for the other variables as well? I just followed the style of the file.

+1 from my side
but maybe wait for other comments

File: kratos/python/add_constitutive_law_to_python.cpp (hunk 1 above)
Comment (philbucher) [reply kratos/python/add_constitutive_law_to_python.cpp]: hm, I would rather go for removing this function for good (since it is deprecated) 
@RiccardoRossi you added this, what do you think?

I am a bit afraid that if we start to do this also others will start doing it, which is against what it is intended to do

After all, it is just a warning, as you write, nothing to concern you if you are not using it
Why is this a problem for you?

File: kratos/utilities/intersection_utilities.h [hunk 4]
```diff
@@ -169,29 +169,29 @@ class IntersectionUtilities
...
+        const TGeometryType& rLineGeometry,
+        const array_1d<double,3>& rLinePoint0,
+        const array_1d<double,3>& rLinePoint1,
         array_1d<double,3>& rIntersectionPoint,
 		const double epsilon = 1e-12) {
 
 		const array_1d<double,3> r = rLineGeometry[1] - rLineGeometry[0];
 		const array_1d<double,3> s = rLinePoint1 - rLinePoint0;
-		const array_1d<double,3> p_q = rLineGeometry[0] - rLinePoint0;		// p - q 
+		//const array_1d<double,3> p_q = rLineGeometry[0] - rLinePoint0;		// p - q
```
Comment (philbucher) [inline kratos/utilities/intersection_utilities.h]: commented

File: kratos/tests/cpp_tests/utilities/test_intersection_utilities.cpp [hunk 5]
```diff
@@ -168,13 +168,13 @@ namespace Testing {
...
-            line_pt_1.Coordinates(), 
-            line_pt_2.Coordinates(), 
+            line_geom,
+            line_pt_1.Coordinates(),
+            line_pt_2.Coordinates(),
             int_pt.Coordinates());
 
         // Compute and check the obtained intersection point coordinates
-        const array_1d<double,3> int_pt_coords = int_pt.Coordinates();
+        //const array_1d<double,3> int_pt_coords = int_pt.Coordinates();
```
Comment (philbucher) [inline kratos/tests/cpp_tests/utilities/test_intersection_utilities.cpp]: commented

File: kratos/tests/cpp_tests/utilities/test_intersection_utilities.cpp [hunk 6]
```diff
@@ -144,13 +144,13 @@ namespace Testing {
...
```
Comment (philbucher) [inline kratos/tests/cpp_tests/utilities/test_intersection_utilities.cpp]: commented

File: kratos/tests/cpp_tests/utilities/test_intersection_utilities.cpp [hunk 7]
```diff
@@ -120,13 +120,13 @@ namespace Testing {
...
```
Comment (philbucher) [inline kratos/tests/cpp_tests/utilities/test_intersection_utilities.cpp]: commented

**Review by philbucher — COMMENTED** [review]
I will remove my block after you removed the commented code

Thanks btw for taking care of this!
Still I am a bit surprised that there are unused variables. Usually Travis won't compile since it turns those warnings into errors

File: kratos/sources/kernel.cpp (hunk 3 above)
Comment (philbucher) [reply kratos/sources/kernel.cpp]: Thanks

File: kratos/python/add_constitutive_law_to_python.cpp (hunk 1 above)
Comment (maceligueta) [reply kratos/python/add_constitutive_law_to_python.cpp]: I discussed this with @RiccardoRossi  yesterday. My point of view is that warnings are not pleasant for anyone who downloads and compiles Kratos, it's something people point at when something goes wrong. Apart from this, any developer concerned about cleaning their code is losing time by checking if this warning is affecting them. A waste of time, in my opinion, if you have nothing to do with that method. So I am in favor of a clean compilation whenever it's possible, and now it's possible.
Removing the call is not possible currently, because the deprecation is too recent and some users might be using this method from Python still.
Anyways, we need to find a stable solution for this kind of problems. I think nobody wants a compilation with several warnings here and there about deprecated methods... 
So, this was my proposal, but any better solution is welcome, of course.

Comment (philbucher) [reply kratos/python/add_constitutive_law_to_python.cpp]: I see, I also don't like warnings :) Yet deprecating things seems to be a better way than breaking things. I think also python/numpy is doing it this way
Not sure if it is possible though to always have a clean core compilation, since a lot of things are there

Different approach:
We could add `-Wno-deprecated-declarations` to the standard-flags, then users won't see those warnings at all
Doesn't change the situation for developers though...

If the code you are adding is necessary, then we could create a macro for it, which can then also be used in similar situations. Ofc a comment has to be added that this macro should rarely be used
~~~c
#define KRATOS_DISABLE_DEPRECATION_WARNING
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
#elif defined(__GNUC__) || defined(__GNUG__)
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
#elif defined(_MSC_VER)
#pragma warning(push)
#pragma warning(disable: 4996)
#endif
~~~

Comment (pooyan-dadvand) [reply kratos/python/add_constitutive_law_to_python.cpp]: I would create two macros having all these definition but showing the intention:

```c++
KRATOS_CALL_DEPRECATED_FUNTION_BEGIN
    rThisConstitutiveLaw.InitializeSolutionStep(rMaterialProperties,
                                                rElementGeometry,
                                                rElementGeometry,
                                                rShapeFunctionsValues,
                                                rShapeFunctionsValues,
                                                rCurrentProcessInfo);
                                                rCurrentProcessInfo);
KRATOS_CALL_DEPRECATED_FUNTION_END
```
And it would be more searchable for future removing

Comment (maceligueta) [reply kratos/python/add_constitutive_law_to_python.cpp]: I thought about this as well, but I didn't find a way to code a macro with pre-processor sentences inside. Maybe I should write all of this in a function? @roigcarlo told me that he could not think of a nice solution in this spot.

Comment (pooyan-dadvand) [reply kratos/python/add_constitutive_law_to_python.cpp]: My idea was something like this but I'm not sure if it works:

```c++
#define KRATOS_DISABLE_DEPRECATION_WARNING
#if defined(__clang__)
#define KRATOS_CALL_DEPRECATED_FUNTION_BEGIN \
_Pragma(clang diagnostic push) \
_Pragma(clang diagnostic ignored "-Wdeprecated-declarations")
#elif defined(__GNUC__) || defined(__GNUG__)
#define KRATOS_CALL_DEPRECATED_FUNTION_BEGIN \
_Pragma(GCC diagnostic push)\
_Pragma(GCC diagnostic ignored "-Wdeprecated-declarations")
#elif defined(_MSC_VER)
#define KRATOS_CALL_DEPRECATED_FUNTION_BEGIN \
__pragma(warning(push))\
__pragma(warning(disable: 4996))
#endif
```

Comment (maceligueta) [reply kratos/python/add_constitutive_law_to_python.cpp]: Thanks @pooyan-dadvand , nice approach! Finally I managed to compile. Note that I needed to create KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION because there are nested literals (_Pragma needs to get an argument between quote marks, but there were two levels of that: "GCC diagnostic ignored "-Wdeprecated-declarations" ").

```c++
#if defined(__clang__)
#define KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(x) _Pragma(#x)
#define KRATOS_START_IGNORING_DEPRECATED_FUNCTION_WARNING \
KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(clang diagnostic push) \
KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(clang diagnostic ignored "-Wdeprecated-declarations")
#elif defined(__GNUC__) || defined(__GNUG__)
#define KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(x) _Pragma(#x)
#define KRATOS_START_IGNORING_DEPRECATED_FUNCTION_WARNING \
KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(GCC diagnostic push) \
KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(GCC diagnostic ignored "-Wdeprecated-declarations")
#elif defined(_MSC_VER)
#define KRATOS_START_IGNORING_DEPRECATED_FUNCTION_WARNING \
__pragma(warning(push))\
__pragma(warning(disable: 4996))
#endif

#if defined(__clang__)
#define KRATOS_STOP_IGNORING_DEPRECATED_FUNCTION_WARNING \
_Pragma("clang diagnostic pop")
#elif defined(__GNUC__) || defined(__GNUG__)
#define KRATOS_STOP_IGNORING_DEPRECATED_FUNCTION_WARNING \
_Pragma("GCC diagnostic pop")
#elif defined(_MSC_VER)
#define KRATOS_STOP_IGNORING_DEPRECATED_FUNCTION_WARNING \
__pragma(warning(pop))
#endif
```

File: kratos/includes/define.h [hunk 8]
```diff
@@ -578,6 +578,34 @@ catch(...) { Block KRATOS_THROW_ERROR(std::runtime_error, "Unknown error", MoreI
 #endif
 
 
+#if defined(__clang__)
```
Comment (philbucher) [inline kratos/includes/define.h:586]: could you please add some comments what these macros are doing? Thx

Comment (maceligueta) [reply kratos/includes/define.h:586]: done!

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h (hunk 2 above)
Comment (maceligueta) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: What's the policy for this @pooyan-dadvand , @RiccardoRossi ?

File: kratos/python/add_constitutive_law_to_python.cpp (hunk 1 above)
Comment (pooyan-dadvand) [reply kratos/python/add_constitutive_law_to_python.cpp]: More complicated than what I would imagine! Good macro name and good job!

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h (hunk 2 above)
Comment (maceligueta) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: ping @pooyan-dadvand @RiccardoRossi . What to do in these cases?

Comment (jcotela) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: I would suggest going ahead and initializing them whenever possible. I guess that the only motive it is not done yet is that this was new in C++11.

**Comment by maceligueta** [comment]
I think I am passing the tests. Actually, I am getting some errors but I'd say they come from the master branch due to a lack of compatibility with Python 2.7. I already started a collaborative PR  in #4045 for fixing them.
If nothing looks wrong, can anyone approve?

**Review by philbucher — APPROVED** [review]
looks good to me

thanks @maceligueta

**Comment by philbucher** [comment]
don't forget to delete the branch :)
